src/libsyntax/json/tests.rs

```diff
@@ -0,0 +1,190 @@
+use super::*;
+
+use crate::tests::{Shared};
+use crate::json::JsonEmitter;
+use crate::source_map::{SourceMap, FilePathMapping};
+use crate::with_default_globals;
+
+use errors::emitter::{HumanReadableErrorType, ColorConfig};
+use errors::Handler;
+use rustc_serialize::json::decode;
+use syntax_pos::{BytePos, Span};
+
+use std::str;
+
+#[derive(RustcDecodable, Debug, PartialEq, Eq)]
+struct TestData {
+    spans: Vec<SpanTestData>,
+}
+
+#[derive(RustcDecodable, Debug, PartialEq, Eq)]
+struct SpanTestData {
+    pub byte_start: u32,
+    pub byte_end: u32,
+    pub line_start: u32,
+    pub column_start: u32,
+    pub line_end: u32,
+    pub column_end: u32,
+}
+
+/// Test the span yields correct positions in JSON.
+fn test_positions(code: &str, span: (u32, u32), expected_output: SpanTestData) {
+
+    let expected_output = TestData { spans: vec![expected_output] };
+
+    with_default_globals(|| {
+        let sm = Lrc::new(SourceMap::new(FilePathMapping::empty()));
+        sm.new_source_file(Path::new("test.rs").to_owned().into(), code.to_owned());
+
+        let output = Arc::new(Mutex::new(Vec::new()));
+        let je = JsonEmitter::new(
+            Box::new(Shared { data: output.clone() }),
+            None,
+            sm,
+            true,
+            HumanReadableErrorType::Short(ColorConfig::Never),
+            false
+        );
+
+        let span = Span::with_root_ctxt(BytePos(span.0), BytePos(span.1));
+        let handler = Handler::with_emitter(true, None, Box::new(je));
+        handler.span_err(span, "foo");
+
+        let bytes = output.lock().unwrap();
+        let actual_output = str::from_utf8(&bytes).unwrap();
+        let actual_output : TestData = decode(actual_output).unwrap();
+
+        println!("expected output:\n------\n{:#?}------", expected_output);
+        println!("actual output:\n------\n{:#?}------", actual_output);
```

Comment:
I am not sure if it's considered good to leave `println`s in the tests. Perhaps this can be moved into the `assert!` call?

```
assert_eq!(expected_output, actual_output, 
"expected output:\n------\n{:#?}------
actual output:\n------\n{:#?}------",
expected_output, actual_output
)
```

Reply:
Shouldn't `assert_eq!` alone do the expected vs actual printout? The current implementation is adapted from `libsyntax/tests.rs` - where `assert!` is used over `assert_eq!`. I'm guessing this might have something to do with `assert_eq!`'s handling of multi-line text.

... although I have no idea where the `{:#?}` format string came from.

I'll move the output to the assert macro, but I'll experiment a bit with `assert!` vs `assert_eq!` to and `{}` vs `{:?}` to see what gives the best output.